### PR TITLE
PR to make patching and stitching agnostic to coordinate direction

### PR DESCRIPTION
**Pull request by MartinSJRogers**
Changes to model.py and loader.py to allow patching during inference, irrespective of whether `x1` and `x2` increase or decrease from the top left corner. In particular, the code is modified to overcome the error deriving from `xr.combine_by_coords` that the coordinates are not monotonic in direction. Summary of changes:

**data/loader.py**

1. Add new method, `self._compute_x1x2_direction()`, to calculate `x1` and `x2` direction
2. Add additional if/else statements to `sample_sliding_window()`, so the top left corner is always the first part of the image to be patched, irrespective of coordinate direction. This can probably be simplified. 

Other minor change in loader.py

1.  sample_sliding_window() sometimes produces duplicate entries, so code added in this method to remove these. 

**model/model.py**

1. Identify direction of `x1` and `x2` in main body of `patch_predict()` (line 985-997)
4. Add `x1_ascend` and `x2_ascend` as arguments to `get_patch_overlap()`, so overlap between patches can be calculated irrespective of coordinate direction
5. Add `x1_ascend` and `x2_ascend` as arguments to `stitch_clipped_predictions()` and add additional if/else statements to make stitching agnostic to x1 and x2 direction

Other minor changes in model.py:

1. Change all references of `x` and `y` to `x2` and `x1` respectively
2. change reference of `unnorm_coords['x1']` to `orig_x1_name`

Prediction when patching (stride = 0.25, patch = 0.5):
image

Prediction when patching (stride = 0.05, patch = 0.1):
image

Prediction when not patching:
image

Difference between prediction when patching/not patching- todo-resolve line!!:
image

### CODE REVIEW

**Comment by nilsleh** [comment]
@MartinSJRogers Thanks for all the work. This looks super promising. A couple of more tests visualizations that might be potentially useful as a sanity check:
- Difference in MSE or other metric between different stride / patch size modes in terms of mean prediction
- Difference in predictive uncertainty between these modes? Would the uncertainty change between modes and non-patching?
- It would be odd if the prediction of the non-patching vs patching is exactly the same right? Maybe just hard to see with the colormap, but I would think that the area besides that stripe is not zero right?

**Comment by MartinSJRogers** [comment]
> @MartinSJRogers Thanks for all the work. This looks super promising. A couple of more tests visualizations that might be potentially useful as a sanity check:
> 
> * Difference in MSE or other metric between different stride / patch size modes in terms of mean prediction
> * Difference in predictive uncertainty between these modes? Would the uncertainty change between modes and non-patching?
> * It would be odd if the prediction of the non-patching vs patching is exactly the same right? Maybe just hard to see with the colormap, but I would think that the area besides that stripe is not zero right?

Thanks for the pointers @nilsleh. To check, by 'mode' do you mean compare the outputs when using different combinations of values for stride and patch size? Good idea and I'll give it a go. I will also do comparative plots of uncertainty. 
As for the values- yes they are near-zero, not zero exactly. I must say, I would have expected the differences to be a little larger, as the size of the input to the UNet component of the model would affect the output- but maybe this will be more apparent when using smaller patch and stride sizes. 
I'll get back when I've played with these ideas!

File: deepsensor/data/loader.py [hunk 1]
```diff
@@ -1413,30 +1460,87 @@ def sample_sliding_window(
...
-                    y0 = x1_max - x1_extend
-                else:
-                    y0 = y
-                if x + x2_extend > x2_max:
-                    x0 = x2_max - x2_extend
-                else:
-                    x0 = x
+        # Todo: simplify these elif statements
+        if self.coord_directions['x1'] == False and self.coord_directions['x2'] == True:
+            for y in np.arange(x1_max, x1_min, -dy):
```
Comment (davidwyld) [inline deepsensor/data/loader.py:1470]:
```
            for y in np.arange(x1_max, x1_min, -dy):
                    y0 = x1_max - x1_extend
                for x in np.arange(x2_min, x2_max, dx):
                else:
                    if y - x1_extend < x1_min:
                    y0 = y
                        y0 = x1_min + x1_extend
                if x + x2_extend > x2_max:
                    else:
                    x0 = x2_max - x2_extend
                        y0 = y
                else:
                    if x + x2_extend > x2_max:
                    x0 = x
                        x0 = x2_max - x2_extend
                    else:
                        x0 = x

                    # bbox of x1_min, x1_max, x2_min, x2_max per patch
                    bbox = [y0 - x1_extend, y0, x0, x0 + x2_extend]
                    patch_list.append(bbox)
```

This sort of block is repeated  4 times here, could we turn this into a reusable function?

File: deepsensor/data/loader.py [hunk 2]
```diff
@@ -1413,30 +1460,87 @@ def sample_sliding_window(
...
+                for x in np.arange(x2_min, x2_max, dx):
+                    if y - x1_extend < x1_min:
```
Comment (davidwyld) [inline deepsensor/data/loader.py:1472]: Is `x1_extend` the intended variable name? Should it be `x1_extent`? (I appreciate that this has been used previously, sorry for just picking up on this now)

File: deepsensor/data/loader.py [hunk 3]
```diff
@@ -1413,30 +1460,87 @@ def sample_sliding_window(
...
+                    bbox = [y0, y0 + x1_extend, x0, x0 + x2_extend]
 
-                # bbox of x1_min, x1_max, x2_min, x2_max per patch
-                bbox = [y0, y0 + x1_extend, x0, x0 + x2_extend]
+                    patch_list.append(bbox)
 
-                patch_list.append(bbox)
+        # Remove duplicate patches while preserving order
+        seen = set()
+        unique_patch_list = []
```
Comment (davidwyld) [inline deepsensor/data/loader.py:1535]: Possibly could leverage the characteristics of `set`s here to create a unique set of patches, but not sure if order is preserved in sets.

File: deepsensor/data/loader.py (hunk 1 above)
Comment (davidwyld) [reply deepsensor/data/loader.py:1470]: Btw I'm keen to get this PR merged so if this is a blocker, we could come back to it.

**Review by davidwyld — APPROVED** [review]
I can't comment too much on the functionality but seems okay to me. I've left a few comments, though they aren't reasons to block merging the PR so I've approved as I'm keen to get this merged in.

**Comment by nilsleh** [comment]
@MartinSJRogers do you have any comments on David's review? Otherwise I can merge this.

**Comment by MartinSJRogers** [comment]
@nilsleh - happy for you to merge. As per our previous conversation, I have been running some additional tests over the weekend to determine the impact of patching on model output. The first set images show the different mean and std values generated when using different patching scenarios. Overall, std reduced when using a smaller patch size (note the different std value ranges), but it identifies greater uncertainty in mountainous regions- which feels appropriate.

image

Below are mean prediction differences when not patching or using a patch size of 0.2 and 0.5. Differences are greater when using a smaller patch size.  
image

Finally I looked at the difference between the patched/non-patched predictions and the test ERA5 data for that date. From the small sample set, errors reduce when using patching. 

image

My main concern is that the training and validation loss are not as stable when patching. The validation loss may be more erratic due to the very different scales on the y-axis, but the training loss is definitely less stable. I think I remember you saying that you generated similar loss curves with/without patching. Was this the case for you?

image

We can discuss this all with Tom, but if you're happy, I'm happy for this to be merged, and we can now start a conversation with Tom in more detail.

**Comment by nilsleh** [comment]
@MartinSJRogers Thanks for all the additional experiments. So overall from those sets of experiments it seems:

- patch-wise predictions during inference can have lower errors than training on the full region
- it also appears that the overall uncertainty is lower with the patching scheme
- train loss is prone to more noise and therefore possibly leads to less stable training

Regarding the last point, intuitively I would expect the loss to be have a "stronger oscillating" behavior than training on a larger region, since the model gets less context. However, there are strategies to mitigate this, should this cause models not to converge, for example gradient accumulation. The experiment I ran was with a patch size of 0.4 and yielded the following plot.

Screenshot from 2024-08-20 20-58-57

So from that it seems that model performance could possibly be a function of patch sized training which is sort of exciting.

Another interesting experiment could be the following: The whole reason why we need patches is because the memory footprint during a training gradient step. However, during inference we do not need gradients and hence it should be possible to fit larger patch sizes during validation and evaluation. This might resolve or reduce some artifacts/issues that could arise from having to do stitching, should that be a problem.

**Comment by MartinSJRogers** [comment]
> @MartinSJRogers Thanks for all the additional experiments. So overall from those sets of experiments it seems:
> 
> * patch-wise predictions during inference can have lower errors than training on the full region
> * it also appears that the overall uncertainty is lower with the patching scheme
> * train loss is prone to more noise and therefore possibly leads to less stable training
> 
> Regarding the last point, intuitively I would expect the loss to be have a "stronger oscillating" behavior than training on a larger region, since the model gets less context. However, there are strategies to mitigate this, should this cause models not to converge, for example gradient accumulation. The experiment I ran was with a patch size of 0.4 and yielded the following plot.
> 
> Screenshot from 2024-08-20 20-58-57
> 
> So from that it seems that model performance could possibly be a function of patch sized training which is sort of exciting.
> 
> Another interesting experiment could be the following: The whole reason why we need patches is because the memory footprint during a training gradient step. However, during inference we do not need gradients and hence it should be possible to fit larger patch sizes during validation and evaluation. This might resolve or reduce some artifacts/issues that could arise from having to do stitching, should that be a problem.

@nilsleh yes I agree with your three summary bullet points. There's always scope for more testing using different datasets, training regimes etc. but these results at least indicate that patching does not produce inferior results. 
Good suggestion regarding gradient accumulation. Is this what you employed when generating the loss curves you shared? If so, would be great if you had a code snippet to share?
Finally, I thought exactly what you did! Yes, patching is required less often during inference. I also ran this experiment where I compared outputs from predict() and predict_patch() on the same model. The difference in outputs is negligible when using a patch size of 0.5. So most change comes from the use of patching during training rather than inference/ stitching. Which is what we would want. 
image

**Comment by nilsleh** [comment]
Didn't do any gradient accumulation there. I'll merge this and we can continue discussing possible experiments/findings with Tom.
